gen/generate.go

```diff
@@ -284,45 +292,69 @@ func generateModule(m *compile.Module, i thriftPackageImporter, builder *generat
 			return nil, err
 		}
 
-		buff := new(bytes.Buffer)
-		if err := g.Write(buff, nil /* fset */); err != nil {
-			return nil, fmt.Errorf(
-				"could not generate idl.go for %q: %v", m.ThriftPath, err)
-		}
+		if len(o.SingleFile) == 0 {
+			buff := new(bytes.Buffer)
+			if err := g.Write(buff, nil); err != nil {
+				return nil, fmt.Errorf(
+					"could not generate idl.go for %q: %v", m.ThriftPath, err)
+			}
 
-		files["idl.go"] = buff.Bytes()
+			files["idl.go"] = buff.Bytes()
+		}
 	}
 
 	// Services must be generated last because names of user-defined types take
 	// precedence over the names we pick for the service types.
 	if len(m.Services) > 0 {
-		for _, serviceName := range sortStringKeys(m.Services) {
-			service := m.Services[serviceName]
-
-			// generateModule gets called only for those modules for which we
-			// need to generate code. With --no-recurse, generateModule is
-			// called only on the root file specified by the user and not its
-			// included modules. Only services defined in these files are
-			// considered root services; plugins will generate code only for
-			// root services, even though they have information about the
-			// whole service tree.
-			if _, err := builder.AddRootService(service); err != nil {
-				return nil, err
-			}
-
-			serviceFiles, err := Service(g, service)
+		if len(o.SingleFile) > 0 {
+			serviceFiles, err := Services(g, m.Services, o.SingleFile)
```

Comment:
From digging a little bit, it looks like we're actually smuggling a bunch of file output into the generator ( when we call `Constant(..)`/`TypeDefinition(..)`/`embedIDL(..)`/ etc. ), and then in this call to `Services(..)`, we suddenly write to a buffer, which writes everything held in the generator into this single file. 

Rather than doing this, `Services(..)` should just make the generator hold a bunch of output, and then write it into a single buffer at then end. Let me know if you need some clarification.

Reply:
Agreed that this is more consistent with the rest of the code in gen/generator.go. I originally implemented my Services logic like this since generating and then writing the code is a pattern of the Service function, which was functionality that I originally tried to emulate.

Breaking out the code templating/generation step and the file writing step does allow me to
share more code in gen/generate.go so I've separated the logic out.